language/move-binary-format/src/check_bounds.rs

```diff
@@ -2,117 +2,141 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::{
+    binary_views::BinaryIndexedView,
     errors::{
         bounds_error, offset_out_of_bounds as offset_out_of_bounds_error, verification_error,
         PartialVMError, PartialVMResult,
     },
     file_format::{
-        Bytecode, CodeOffset, CompiledModuleMut, Constant, FieldHandle, FieldInstantiation,
-        FunctionDefinition, FunctionDefinitionIndex, FunctionHandle, FunctionInstantiation,
-        ModuleHandle, Signature, SignatureToken, StructDefInstantiation, StructDefinition,
-        StructFieldInformation, StructHandle, TableIndex,
+        AbilitySet, Bytecode, CodeOffset, CodeUnit, CompiledModule, CompiledScript, Constant,
+        FieldHandle, FieldInstantiation, FunctionDefinition, FunctionDefinitionIndex,
+        FunctionHandle, FunctionInstantiation, ModuleHandle, Signature, SignatureToken,
+        StructDefInstantiation, StructDefinition, StructFieldInformation, StructHandle, TableIndex,
     },
     internals::ModuleIndex,
     IndexKind,
 };
 use move_core_types::vm_status::StatusCode;
 use std::u8;
 
+enum BoundsCheckingContext {
+    Module,
+    ModuleFunction(FunctionDefinitionIndex),
+    Script,
+}
 pub struct BoundsChecker<'a> {
-    module: &'a CompiledModuleMut,
-    current_function: Option<FunctionDefinitionIndex>,
+    view: BinaryIndexedView<'a>,
+    context: BoundsCheckingContext,
 }
 
 impl<'a> BoundsChecker<'a> {
-    pub fn verify(module: &'a CompiledModuleMut) -> PartialVMResult<()> {
+    pub fn verify_script(script: &'a CompiledScript) -> PartialVMResult<()> {
+        let mut bounds_check = Self {
+            view: BinaryIndexedView::Script(&script),
+            context: BoundsCheckingContext::Script,
+        };
+        bounds_check.verify_impl()?;
+        // The bounds checker has already checked each function definition's code, but a script's
+        // code exists outside of any function definition. It gets checked here.
```

Comment:
I think we should check if `self.parameters` is out of bounds here before accessing the signature. I wonder if it would still panic if we get this fixed?

Reply:
Thanks for the suggestion and the explanations! I actually did have some code stashed that checks whether the given parameters index is out of bounds, but I wasn't sure it was correct. I uploaded a patch along the lines of what you suggested (I just need to add a regression test).